### ISSUE TITLE
Directed graph: edges_undirected()?

### ISSUE THREAD

**Issue by mankinskin** [petgraph/petgraph#339]
The [edges()](https://docs.rs/petgraph/0.5.0/petgraph/graph/struct.Graph.html?search=#method.edges) method of Graph returns all outgoing edges of a node if the graph uses Directed edges. How to get all edges of a node in a directed Graph?

There is neighbors_undirected, so it would make sense to provide an `edges_undirected()` method.

**Comment by MiltFra** [comment]
I've had a look into it. It appears, at least from my understanding, that implementing this using the current `Edges` type does not seem feasible. If it were (and it might be, I've only started reading the code today) adding this might be as simple as 

```
pub fn edges_undirected(&self, a: NodeIndex<Ix>) -> Edges<E, Ty, Ix> {
        if !Ty::is_directed() {

            return self.edges_directed(a, Outgoing);
        }
        let mut incoming = self.edges_directed(a, Incoming);
        let mut outgoing = self.edges_directed(a, Outgoing);
        incoming.chain(outgoing).collect()
    }
```

The core issue here is obviously the `.collect()` as Edges currently does not implement `FromIterator` for `IntoIterator<EdgeReference<'a, E, Ix>>`. Sadly, this seems basically impossible to do without changing the type itself drastically. This is due to the fact that `Edges::next()` makes use of the already present links between the edges in a `Graph`. Using the current `next()` method would require changing the `next` value and therefore altering the `Edges` which should not be done under any circumstances.

This leads to the question: How do we implement this? Of course one could use the very same function, just with a different signature. This does not seem very clean and would require API changes when the implementation changes. 

Therefore I propose to use a type `EdgesUndirected` that is either an alias to `Chain<Edges<E, Ty, Ix>>` or a wrapper around it. Please let me know which version you prefer and I'll implement it tomorrow. If you have another idea or any criticism of my code above please tell me, too.

**Comment by mankinskin** [comment]
Yes I have had the same issue when writing my wrapper library, and I ended up returning `Vec`s everywhere. It would be better to return iterators obviously, but I am unsure what the best implementation would be, as I am not experienced in this code base. I may look into it some time, but I am probably fine with any implementation anyone else comes up with.

**Comment by MiltFra** [comment]
So, I've been tinkering around with this and I think I've now found a decent solution. It requires changing the `Edges` type but it doesn't break any current API and it makes the `edges_undirected()` signature look as you'd expected it to.

Here's what I did. I renamed the current `Edges` into `EdgeIter` (The name may be discussed.) and created a new `Edges` type which looks as follows:
```rust
#[derive(Clone)]
pub enum Edges<'a, E: 'a, Ty, Ix: 'a = DefaultIx>
where
    Ty: EdgeType,
    Ix: IndexType,
{
    Single(EdgeIter<'a, E, Ty, Ix>),
    Chain(Chain<EdgeIter<'a, E, Ty, Ix>, EdgeIter<'a, E, Ty, Ix>>),
}

impl<'a, E, Ty, Ix> Iterator for Edges<'a, E, Ty, Ix>
where
    Ty: EdgeType,
    Ix: IndexType,
{
    type Item = EdgeReference<'a, E, Ix>;

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            Edges::Single(iter) => iter.next(),
            Edges::Chain(iter) => iter.next(),
        }
    }
}
```

The `edges_*` methods now looks as follows:
```rust
   pub fn edges_directed(&self, a: NodeIndex<Ix>, dir: Direction) -> Edges<E, Ty, Ix> {
        Edges::Single(self.edges_directed_iter(a, dir))
    }

    #[inline]
    fn edges_directed_iter(&self, a: NodeIndex<Ix>, dir: Direction) -> EdgeIter<E, Ty, Ix> {
        EdgeIter {
            skip_start: a,
            edges: &self.edges,
            direction: dir,
            next: match self.nodes.get(a.index()) {
                None => [EdgeIndex::end(), EdgeIndex::end()],
                Some(n) => n.next,
            },
            ty: PhantomData,
        }
    }

    pub fn edges_undirected(&self, a: NodeIndex<Ix>) -> Edges<E, Ty, Ix> {
        if Ty::is_directed() {
            Edges::Chain(
                self.edges_directed_iter(a, Outgoing)
                    .chain(self.edges_directed_iter(a, Incoming)),
            )
        } else {
            self.edges_directed(a, Outgoing)
        }
    }
```

Why did I not just wrap `Chain<..., ...> ` into a new type? Well, the issue is that, depending on the type of the `Graph`, `edges_undirected(a)` might just refer to `edges_directed(a, Outgoing)`. This means that their return types either need to be identical or I'd need to do something dirty by chaining the result with an empty iterator which I think is a very bad solution. 

The fact that `Edges` is an enum with two variants now has the potential drawbacks that it has gotten slower overall. Until this has been investigated further I can't really say how bad this is. Since you seem to be working with the library it would be nice if you could run some of your own benchmarks with the current release of the library and compare them to my [fork](https://github.com/miltfra/petgraph).

Another option, which I can't imagine is any faster, might be to make `Edges` a wrapper for a generic iterator over `EdgeReference`. This would require to box the original iterator, though.

**Comment by mankinskin** [comment]
Looks good, I can benchmark your implementation after I finished my current objective :)

**Comment by MiltFra** [comment]
I used the existing Dijkstra benchmark with different `NODE_COUNT`s, here's what I found:

Without `edges_undirected`:
```
test dijkstra_1000    ... bench:     187,703 ns/iter (+/- 22,996)
test dijkstra_10000   ... bench:   1,791,813 ns/iter (+/- 25,205)
test dijkstra_100000  ... bench:  20,502,777 ns/iter (+/- 5,839,090)
test dijkstra_1000000 ... bench: 245,786,340 ns/iter (+/- 14,762,037)
```

With `edges_undirected`:
```
test dijkstra_1000    ... bench:     187,703 ns/iter (+/- 22,996)
test dijkstra_10000   ... bench:   1,791,813 ns/iter (+/- 25,205)
test dijkstra_100000  ... bench:  20,502,777 ns/iter (+/- 5,839,090)
test dijkstra_1000000 ... bench: 245,786,340 ns/iter (+/- 14,762,037)
```

So the Dijkstra in my implementation is about 10% slower. I don't think that should be the case, given that we're trying to implement a new feature and not fix a bug.

What do you think?

**Comment by antonok-edm** [comment]
@MiltFra I encourage you to open a PR with this change, it already looks like it could be useful to a number of people and I think you're likely to get better feedback on the implementation in a PR.

**Comment by mankinskin** [comment]
Sorry, I could not really prioritize this after all.
